### PR TITLE
[APM-697] Improve collect, aggregate, etc performance

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Mostly via avoiding copies big values (byte size > 15) from rocksdb
Part of APM-697 task

TODO list for next PRs:
1. Replace malloc/free via new/delete for vpack buffer
2. Remove special case for int48
3. Fix projection behavior
4. ...

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1385
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Graph/Cache/RefactoredTraverserCache.cpp [hunk 1]
```diff
@@ -120,75 +120,66 @@ bool RefactoredTraverserCache::appendEdge(EdgeDocumentToken const& idToken,
...
+            transaction::BuilderLeaser builder(_trx);
+            {
+              VPackObjectBuilder guard(builder.get());
+              _edgeProjections.toVelocyPackFromDocument(*builder, edge, _trx);
+            }
+            result = aql::AqlValue(builder->slice());
+          } else {
+            result = aql::AqlValue(edge);
+          }
+          result = aql::AqlValue(edge);
```
Comment (jsteemann) [inline arangod/Graph/Cache/RefactoredTraverserCache.cpp:170]: I see that this is identical to the original code, but it seems that we are unconditionally clobbering `result` in line 170, even if it was set to something else in lines 166 or 168. This looks inefficient and wrong. Especially it looks like that the projections currently cause _extra_ work, and the full edge is copied out later anyway.

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 2]
```diff
@@ -2021,24 +1876,30 @@ Result RocksDBCollection::lookupDocumentVPack(
...
     // write entry back to cache
     cache::Cache::SimpleInserter<DocumentCacheType>{
         static_cast<DocumentCacheType&>(*cache), key->string().data(),
         static_cast<uint32_t>(key->string().size()), ps.data(),
         static_cast<uint64_t>(ps.size())};
   }
 
+  auto data = buffer.release();
+  cb(token, std::move(data));
+  buffer.acquire(std::move(data));
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1901]: `data` is moved twice here. While this may work (if `cb` does not steal the value from `data`), it looks strange to the reader of the code.

File: arangod/RocksDBEngine/RocksDBIncrementalSync.cpp [hunk 3]
```diff
@@ -342,9 +345,9 @@ Result syncChunkRocksDB(DatabaseInitialSyncer& syncer,
           auto [documentId, revisionId] = lookupResult;
 
           tempBuilder->clear();
-          r = physical->lookupDocument(*trx, documentId, *tempBuilder,
-                                       /*readCache*/ true, /*fillCache*/ false,
-                                       ReadOwnWrites::yes);
+          auto r =
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBIncrementalSync.cpp:348]: This assigns the lookup result to a new in-scope variable `r`, whereas the old code assigned the lookup result to a variable `r` from an outer scope.
Is this change intentional?

File: arangod/RocksDBEngine/RocksDBIncrementalSync.cpp [hunk 4]
```diff
@@ -889,9 +887,10 @@ Result handleSyncKeysRocksDB(DatabaseInitialSyncer& syncer,
               auto [documentId, revisionId] = lookupResult;
 
               tempBuilder.clear();
-              r = physical->lookupDocument(
-                  *trx, documentId, tempBuilder, /*readCache*/ true,
-                  /*fillCache*/ false, ReadOwnWrites::yes);
+              r = physical->lookup(trx.get(), documentId, callback,
+                                   {.readCache = true,
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBIncrementalSync.cpp:891]: nit: `.readCache = true` seems to be the default and doesn't need to be mentioned explicitly here.
All previous callers of `lookup(...)` at least did not specify it explicity, so this place here is a little inconsistent with the previous occurrences.

File: arangod/Aql/AqlValue.cpp [hunk 5]
```diff
@@ -81,302 +83,209 @@ static inline uint8_t intLength(int64_t value) noexcept {
...
-    case RANGE: {
-      break;
-    }
+    case VPACK_INLINE:
+      return VPackSlice{_data.inlineSliceMeta.slice}.isNumber();
+    case VPACK_SLICE_POINTER:
+      return VPackSlice{_data.slicePointerMeta.pointer}.isNumber();
+    case VPACK_MANAGED_SLICE:
+      return VPackSlice{_data.managedSliceMeta.pointer}.isNumber();
+    default:
```
Comment (jsteemann) [inline arangod/Aql/AqlValue.cpp:188]: Is it somehow possible that we ever attempt to store a number in a `VPACK_MANAGED_STRING` AqlValue? Then this case would need to be handled here too.

File: arangod/RocksDBEngine/RocksDBIncrementalSync.cpp (hunk 3 above)
Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBIncrementalSync.cpp:348]: No, it's not, thanks!

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 2 above)
Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBCollection.cpp:1901]: Changed

File: arangod/Aql/AqlValue.cpp (hunk 5 above)
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:188]: No, it's not, you can see it in ctor

File: arangod/Graph/Cache/RefactoredTraverserCache.cpp (hunk 1 above)
Comment (MBkkt) [reply arangod/Graph/Cache/RefactoredTraverserCache.cpp:170]: I just comment out code above, if you know how to fix it, feel free to suggest

File: arangod/Aql/AqlValue.cpp [hunk 6]
```diff
@@ -1339,64 +1068,52 @@ int AqlValue::Compare(velocypack::Options const* options, AqlValue const& left,
...
-AqlValue::AqlValue() noexcept {
-  // construct a slice of type None
-  // we will simply zero-initialize the two 64 bit words
-  _data.words[0] = 0;
-  _data.words[1] = 0;
-
-  // VPACK_INLINE must have a value of 0, and VPackSlice::None must be equal
-  // to a NUL byte too
-  static_assert(AqlValueType::VPACK_INLINE == 0,
-                "invalid value for VPACK_INLINE");
```
Comment (jsteemann) [inline arangod/Aql/AqlValue.cpp:1356]: Is this `static_assert` kept somewhere (probably in the header file)?
It is important that a default constructed AqlValue has type VPACK_INLINE and contains the velocypack None type. This is ensure by making all bits in the AqlValue struct zero. Would be good to keep the assertion somewhere.

File: arangod/Aql/AqlValue.cpp [hunk 7]
```diff
@@ -1339,64 +1068,52 @@ int AqlValue::Compare(velocypack::Options const* options, AqlValue const& left,
...
+  setType(t);
+  switch (t) {
     case VPACK_MANAGED_SLICE:
       _data.managedSliceMeta.lengthOrigin =
           other._data.managedSliceMeta.lengthOrigin;
-      _data.managedSliceMeta.managedPointer = static_cast<uint8_t*>(data);
+      _data.managedSliceMeta.pointer =
+          const_cast<uint8_t*>(static_cast<uint8_t const*>(data));
       break;
-    case VPACK_SLICE_POINTER:
```
Comment (jsteemann) [inline arangod/Aql/AqlValue.cpp:1388]: Is this case removed intentionally? Why?

File: arangod/Aql/AqlValue.cpp [hunk 8]
```diff
@@ -1339,64 +1068,52 @@ int AqlValue::Compare(velocypack::Options const* options, AqlValue const& left,
...
+  if (size < sizeof(AqlValue)) {
+    initFromSlice(slice, size);
+  } else {
+    setType(AqlValueType::VPACK_MANAGED_STRING);
+    _data.managedStringMeta.pointer = data.release();
+  }
 }
 
 AqlValue::AqlValue(uint8_t const* pointer) noexcept {
+  // TODO(MBkkt) Is external possible here?
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:1093]: It wasn't in your PR, and it's affect only this ctor, so I added todo to check it later

File: arangod/Aql/AqlValue.cpp (hunk 7 above)
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:1388]: Because this function should be called only for destructible AqlValue.
So it looks like "typo" in old code

File: arangod/Aql/AqlValue.cpp (hunk 6 above)
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:1356]: It's kept in erase function, which called from dtor now

File: arangod/Aql/AqlValue.cpp (hunk 7 above)
Comment (jsteemann) [reply arangod/Aql/AqlValue.cpp:1388]: ok

File: arangod/Aql/AqlValue.cpp [hunk 9]
```diff
@@ -1633,46 +1346,45 @@ Range const* AqlValue::range() const {
...
-    case VPACK_INLINE_INT48:
-    case VPACK_INLINE_INT64:
-    case VPACK_INLINE_UINT64:
-    case VPACK_INLINE_DOUBLE:
-    case VPACK_SLICE_POINTER:
-      return 0;
     case VPACK_MANAGED_SLICE:
       return _data.managedSliceMeta.getLength();
+    case VPACK_MANAGED_STRING:
+      return sizeof(std::string) + _data.managedStringMeta.getLength();
```
Comment (jsteemann) [inline arangod/Aql/AqlValue.cpp:1368]: Nit: this will slightly underreport the memory usage, as we are at least missing the one byte for the null terminator. Instead of adding the string's `size()` here, we could also use its `capacity()`, because the string allocation may be larger than what it stores.

Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp]: Fixed

File: arangod/Aql/AqlValue.cpp (hunk 7 above)
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:1388]: You can also see assert(false) in default case, so it also will be checked by tests.

File: arangod/Graph/Cache/RefactoredTraverserCache.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Graph/Cache/RefactoredTraverserCache.cpp]: We can leave it as you did it now in this PR, and fix it later when this PR is merged.

File: arangod/Aql/AqlValue.cpp (hunk 7 above)
Comment (MBkkt) [reply arangod/Aql/AqlValue.cpp:1388]: Same for `AqlValue::data()`

**Review by jsteemann — APPROVED** [review]
LGTM
